pkg/kubectl/cmd/get.go

```diff
@@ -48,66 +51,136 @@ Examples:
+	if generic {
+		// the outermost object will be converted to the output-version
+		printer = kubectl.NewVersionedPrinter(printer, api.Scheme, outputVersion(cmd))
 
-			// print the current object
-			if !isWatchOnly {
-				if err := printer.PrintObj(obj, out); err != nil {
-					checkErr(fmt.Errorf("unable to output the provided object: %v", err))
-				}
-			}
+		obj, err := b.Flatten().Do().Object()
+		checkErr(err)
 
-			// print watched changes
-			if isWatch || isWatchOnly {
-				rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
-				checkErr(err)
+		err = printer.PrintObj(obj, out)
+		checkErr(err)
+		return
+	}
 
-				w, err := restHelper.Watch(namespace, rv, labelSelector, labels.Everything())
-				checkErr(err)
+	// use the default printer for each object
+	err = b.Do().Visit(func(r *resource.Info) error {
+		printer, err := printerForMapping(f, cmd, r.Mapping)
+		if err != nil {
+			return err
+		}
+		return printer.PrintObj(r.Object, out)
+	})
+	checkErr(err)
+}
 
-				kubectl.WatchLoop(w, printer, out)
-			}
-		},
+// outputVersion returns the preferred output version for generic content (JSON, YAML, or templates)
+func outputVersion(cmd *cobra.Command) string {
+	outputVersion := GetFlagString(cmd, "output-version")
+	if len(outputVersion) == 0 {
+		outputVersion = GetFlagString(cmd, "api-version")
 	}
-	cmd.Flags().StringP("output", "o", "", "Output format: json|yaml|template|templatefile")
-	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version)")
-	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers")
-	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when --output=template or --output=templatefile")
-	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
-	cmd.Flags().BoolP("watch", "w", false, "After listing/getting the requested object, watch for changes.")
-	cmd.Flags().Bool("watch-only", false, "Watch for changes to the requseted object(s), without listing/getting first.")
-	return cmd
+	return outputVersion
+}
+
+// printerForCommand returns the default printer for this command.
+func printerForCommand(cmd *cobra.Command) (kubectl.ResourcePrinter, bool, error) {
+	outputFormat := GetFlagString(cmd, "output")
+	templateFile := GetFlagString(cmd, "template")
+	if len(outputFormat) == 0 && len(templateFile) != 0 {
+		outputFormat = "template"
+	}
+
+	return kubectl.GetPrinter(outputFormat, templateFile)
+}
+
+// printerForMapping returns a printer suitable for displaying the provided resource type.
+func printerForMapping(f *Factory, cmd *cobra.Command, mapping *meta.RESTMapping) (kubectl.ResourcePrinter, error) {
+	printer, ok, err := printerForCommand(cmd)
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		printer = kubectl.NewVersionedPrinter(printer, mapping.ObjectConvertor, outputVersion(cmd))
```

Comment:
@lavalamp note I moved versioning to a wrapped printer so that I could precisely control versioning here.  You may want to take a look at the watch behavior and see whether it still makes sense.